bridgesync/downloader.go

```diff
@@ -377,46 +402,68 @@ func findCall(rootCall call, targetAddr common.Address, callback func(call) (boo
 
-// extractCall tries to extract the call for the transaction identified by transaction hash.
-// It relies on debug_traceTransaction JSON RPC function.
-func extractCall(client aggkittypes.RPCClienter, contractAddr common.Address, txHash common.Hash, logger *logger.Logger,
-) (*call, error) {
-	c := &call{To: contractAddr}
-	err := client.Call(c, debugTraceTxEndpoint, txHash, tracerCfg{Tracer: callTracerType})
+// extractRootCall extracts the root call for a transaction using debug_traceTransaction.
+func extractRootCall(client aggkittypes.RPCClienter, contractAddr common.Address, txHash common.Hash) (*call, error) {
+	rootCall := &call{To: contractAddr}
+	err := client.Call(rootCall, debugTraceTxEndpoint, txHash, tracerCfg{Tracer: callTracerType})
 	if err != nil {
 		return nil, err
 	}
-
-	return findCall(*c, contractAddr, nil, logger)
+	return rootCall, nil
 }
 
-// setClaimCalldata traces the transaction to find and decode calldata for the given bridge address.
-//
+// extractCallData extracts both the found call and optionally the root call for a transaction.
+// This function combines extractRootCall and findCall into a single operation for better performance.
 // Parameters:
-// - client: RPC client to fetch the transaction trace.
-// - bridge: Target contract address.
-// - txHash: Transaction hash to trace.
-// - logger: Logger instance for debug logging.
-//
-// Returns an error if tracing fails or calldata isn't found.
-func (c *Claim) setClaimCalldata(
+// - client: RPC client for making debug_traceTransaction calls
+// - bridgeAddr: Target bridge contract address
+// - txHash: Transaction hash to trace
+// - needTxnSender: Whether the root call (for txn_sender) is needed
+// - logger: Logger instance for debug logging
+// Returns:
+// - foundCall: The specific call to the bridge contract (for calldata)
+// - rootCall: The root call (only if needTxnSender is true, otherwise nil)
+// - error: Any error that occurred during extraction
+func extractCallData(
 	client aggkittypes.RPCClienter,
-	bridge common.Address,
+	bridgeAddr common.Address,
 	txHash common.Hash,
+	needTxnSender bool,
 	logger *logger.Logger,
-) error {
-	callFrame := &call{}
-	err := client.Call(callFrame, debugTraceTxEndpoint, txHash, tracerCfg{Tracer: callTracerType})
+) (*call, *call, error) {
+	// Extract root call first
+	rootCall, err := extractRootCall(client, bridgeAddr, txHash)
 	if err != nil {
-		return err
+		return nil, nil, err
 	}
 
-	// Check if the root call was successful
-	if callFrame.Err != nil {
-		return fmt.Errorf("execution reverted in root call (block %d, tx %s): %s", c.BlockNum, txHash.Hex(), *callFrame.Err)
+	// Find the specific call to the bridge contract
+	foundCall, err := findCall(*rootCall, bridgeAddr, nil, logger)
+	if err != nil {
+		return nil, nil, err
 	}
 
-	_, err = findCall(*callFrame, bridge,
+	// Return root call only if needed
+	if needTxnSender {
+		return foundCall, rootCall, nil
+	}
+	return foundCall, nil, nil
+}
```

Comment:
IMO I'd return a `rootCall` anyway. 

Also, as a minor further improvement in terms of readability we could consider creating a struct containing `RootCall` and `FoundCall` fields, as that way it would be clearer from the caller perspective which one is the `rootCall` and what is the `foundCall` one. WDYT?

```suggestion
func extractCallData(
	client aggkittypes.RPCClienter,
	bridgeAddr common.Address,
	txHash common.Hash,
	logger *logger.Logger,
) (*call, *call, error) {
	// Extract root call first
	rootCall, err := extractRootCall(client, bridgeAddr, txHash)
	if err != nil {
		return nil, nil, err
	}

	// Find the specific call to the bridge contract
	foundCall, err := findCall(*rootCall, bridgeAddr, nil, logger)
	if err != nil {
		return nil, nil, err
	}

	return foundCall, rootCall, nil
}
```